src/team_lib/mhba/xccl_mhba_collective.c

```diff
@@ -414,28 +439,53 @@ xccl_mhba_send_blocks_start_with_transpose(xccl_coll_task_t *task)
                 transpose_square_mat(request->transpose_buf_mr->addr,
                                      block_size, request->args.buffer_info.len,
                                      request->tmp_transpose_buf);
-                status = send_block_data(
-                    team->net.qps[cyc_rank],
-                    (uintptr_t)request->transpose_buf_mr->addr, block_msgsize,
-                    request->transpose_buf_mr->lkey, remote_addr,
-                    team->net.rkeys[cyc_rank], IBV_SEND_SIGNALED, 0);
-                if (status != XCCL_OK) {
-                    xccl_mhba_error("Failed sending block [%d,%d,%d]", i, j, k);
-                    return status;
+                if(team->is_dc){
+                    ibv_wr_start(team->net.dcis[current_dci].dc_qpex);
```

Comment:
why don't put ibv_wr_start and complete into send_block_data_dc and make that function return xccl_status_t - same as send block_data_rc. Then you can reduce error handling code here:
if (team->is_dc) {
status = send_block_data_dc
} else {
status = send_block_data_rc
}
if (XCCL_OK != status) {
error
}

Reply:
because the meaning of start and complete is the actual posing to the queue - i checked, and posting all WR at once is a bit faster than posting each WR one by one. so that is done on purpose - that way i'm posting all at once per destination node, meaning per QP.